### PR TITLE
Recipe Checks

### PR DESCRIPTION
**Pull request by mravanelli**
With this PR, I would like to add some checks on the recipes. The tests will be based on the content of the `recipes.csv`, which we recently created to monitor all the recipes and their connected resources (HF, gdrive, etc). Feel free to contribute with some ideas if you want. Only with more testing, we can scale up SpeechBrain significantly as planned.

To do:
- [x] - Check  if all the yaml files are listed in `recipes.csv`
- [x] - Remove DNS recipe (not maintained anymore)
- [x] - Check if all the files reported in recipe.csv exist
- [x] - Make sure all the recipes have a README file, an hparam file, and a train script.
- [x] - Make sure that gdrive and HF repositories are mentioned in the README file (if any)
- [x] - Write routine to run recipes tests based the field `test_debug_flags` and test_debug_checks of

### CODE REVIEW

**Comment by mravanelli** [comment]
With the latest commits, I added a library for supporting recipe tests. It works in this way:
1. From the centralized recipe files (`tests/recipes.csv`) it runs a test using the flags available in the field `test_debug_flags`.
2. We check the return status and raises an error it is !=0
3. We check if all the expected files have been created. The expected files are listed in the field test_debug_checks (e.g. `file_exists=[file1.log, file2.txt,...]`)

One can manually run the recipe tests by running (from the main speechbrain folder):
`tests/.run-recipe-tests.sh
`
For now, I only added one test (`TIMIT/ASR/CTC`), but we might need to add many more. To add a new test, we need to write proper flags in the field t`est_debug_flags` of `tests/recipes.csv`. This is the most annoying part, but one can take inspiration from the test already done or `tests/.run-recipe-tests-old.sh`. The idea is to reuse as much as possible the samples available in tests/samples/ and the annotation (CSV or JSON files) in tests/samples/annotation/. In some cases, the recipe might need some minor changes (e,g., add a flag `skip_preparation` to skip data preparation).

What remains to do, is to find a way to check some values in the loggers. For instance, we might want to check if the `train_loss` is smaller than a threshold in the `train_log.txt`.  I think we need to introduce some specific  syntax for this check like `check_theshold=['train_log.txt', pattern, value]`. Any idea is welcome.

Please @anautsch , double-check it and share your comments/suggestions!

File: speechbrain/utils/recipe_tests.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+
+def check_performance(
+    check_str, output_folder, recipe_id, pattern=r"performance_check=\[(.*?)\]"
+):
+    """Checks if the performance achieved by the recipe matches with the
+    expectations. This is done by adding a performance_check entry in the recipe
+    check field of the csv recipe file
+    For instance: performance_check=[train_log.txt, train loss, <=15, epoch: 2]),
+    will check the variable "train_loss" in the train_log.txt at epoch 2. It will
+    raise an error if the train_loss is <=15.
```
Comment (pplantinga) [inline speechbrain/utils/recipe_tests.py:158]: Should this read "raise an error if the train_loss is > 15" ?
```suggestion
    raise an error if the train_loss is >15.
```

File: speechbrain/utils/recipe_tests.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+    ---------
+    rc: bool
+        The return code obtained after running the command. If 0, the test is
+        run without errors. If >0 the execution failed.
+    """
+    try:
+        os.makedirs(folder)
+    except OSError as e:
+        if e.errno != errno.EEXIST:
+            raise
```
Comment (pplantinga) [inline speechbrain/utils/recipe_tests.py:334]: Is this function necessary? I think `os.makedirs` already includes an option called `exist_ok` which ignores if folder already exists.

File: speechbrain/utils/recipe_tests.py [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+    check = True
+    for i, recipe_id in enumerate(test_script.keys()):
+        print(
+            "(%i/%i) Running test for %s..."
+            % (i + 1, len(test_script.keys()), recipe_id)
+        )
+
+        output_fold = os.path.join(
+            output_folder, recipe_id
+        )  # Remove folder from the last run
```
Comment (pplantinga) [inline speechbrain/utils/recipe_tests.py:408]: Does this code actually remove folder from the last run? It doesn't look like it does to me.

File: speechbrain/utils/recipe_tests.py [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+                % (recipe_id, stderr_file, stdout_file)
+            )
+            check = False
+
+        # Checks
+        check_str = test_check[recipe_id].strip()
+        if do_checks and len(check_str) > 0:
+
+            # Check if the expected files exist
+            check_outcome = check_files(check_str, output_fold, recipe_id)
```
Comment (pplantinga) [inline speechbrain/utils/recipe_tests.py:443]:
```suggestion
            check &= check_files(check_str, output_fold, recipe_id)
```

File: speechbrain/utils/recipe_tests.py [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+
+            # Additional checks might be added here
+            if not (check_outcome):
+                check = False
+
+            check_outcome = check_performance(check_str, output_fold, recipe_id)
+            if not (check_outcome):
+                check = False
```
Comment (pplantinga) [inline speechbrain/utils/recipe_tests.py:451]:
```suggestion
            check &= check_performance(check_str, output_fold, recipe_id)
```

**Review by pplantinga — COMMENTED** [review]
I think these checks will generally keep recipes from some basic mistakes, so its probably a good addition to the toolkit.

At the risk of repeating myself, this change seems to institute a process by which we add new recipes, and I'm a little worried that we're making more work for ourselves than necessary by trying to keep all of these recipes up-to-date. What seems better to me is to consistently maintain a few critical high-quality recipes and for the rest of the recipes we could pin a version to them and let someone else update the recipe if they need it.

**Comment by mravanelli** [comment]
Hi @pplantinga, thank you for the comments! Maintaining several recipes is hard, but we can do our best to make it possible (these tests go in this direction). I think it is super important to have several recipes in the toolkit as this makes research faster. This is why other popular toolkits such as ESPNET (see https://github.com/espnet/espnet/tree/master/egs2) or Kaldi have a long list of recipes as well.

File: tests/.run-recipe-tests-old.sh [hunk 6]
```diff
@@ -0,0 +1,97 @@
+#!/bin/bash
+# This script runs several recipes in debug modality with the small dataset saved in samples.
+# It must be called from the main SpeechBrain folder:
+# tests/.run-recipe-tests.sh
+#
+# Author: Mirco Ravanelli 2021
```
[the remark this answers is not in the document]
Comment (mravanelli) [reply tests/.run-recipe-tests-old.sh:6]: No, I will delete it. But it contains the test flags that I have to copy and paste in the recipes.csv.

File: tests/consistency/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# How to add your recipe in tests/recipes.csv
```
[the remark this answers is not in the document]
Comment (mravanelli) [reply tests/consistency/README.md:1]: This file is linked when errors are detected. However, we can definitely write a guide about how to write a recipe and put it in the website.

File: tests/samples/annotation/ASR_dev.json [hunk 8]
```diff
@@ -1,18 +1,18 @@
 {
     "spk1_snt5": {
         "wav": "{data_root}/spk1_snt5.wav",
-        "length": 2.6,
+        "duration": "2.6",
```
[the remark this answers is not in the document]
Comment (mravanelli) [reply tests/samples/annotation/ASR_dev.json:4]: Yes, I have an error with some recipes without that...

File: speechbrain/utils/recipe_tests.py [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+ * Mirco Ravanelli 2022
+"""
+import os
+import re
+import csv
+import errno
+import subprocess as sp
+
+
+def do_test(row, filter_fields, filters, test_field):
```
[the remark this answers is not in the document]
Comment (mravanelli) [reply speechbrain/utils/recipe_tests.py:13]: Sure, I change change it in my next commit!

File: speechbrain/utils/recipe_tests.py (hunk 3 above)
Comment (mravanelli) [reply speechbrain/utils/recipe_tests.py:408]: No, this is something we can avoid as it is not really necessary.

**Comment by mravanelli** [comment]
ok, everything looks fine now.  I only need help to set up more recipe tests (we can do that with the help of @anautsch in the future). We can merge it for now anyway.

**Pull request merged**